File: data-pipeline/de-normalization/src/main/java/org/ekstep/ep/samza/task/DeNormalizationConfig.java
package org.ekstep.ep.samza.task;


import org.apache.samza.config.Config;

import java.util.ArrayList;
import java.util.List;

public class DeNormalizationConfig {

    private final String JOB_NAME = "DeNormalization";
    private static final String deviceDataJobFlag = "device_data_retrieved";
    private static final String userDataJobFlag = "user_data_retrieved";
    private static final String contentDataJobFlag = "content_data_retrieved";
    private static final String dialCodeDataJobFlag = "dialcode_data_retrieved";
    private static final String collectionDataJobFlag = "collection_data_retrieved";
    private static String authorizationKey = "";

    private String successTopic;
    private String failedTopic;
    private String malformedTopic;
    private Integer ignorePeriodInMonths;
    private List<String> summaryFilterEvents;
    private String apiHost;
    private String dialCodeAPiEndPoint;


    public DeNormalizationConfig(Config config) {
        successTopic = config.get("output.success.topic.name", "telemetry.denorm");
        failedTopic = config.get("output.failed.topic.name", "telemetry.failed");
        malformedTopic = config.get("output.malformed.topic.name", "telemetry.malformed");
        ignorePeriodInMonths = config.getInt("telemetry.ignore.period.months", 6);
        List<String> defaultSummaryEvents = new ArrayList<String>();
        defaultSummaryEvents.add("ME_WORKFLOW_SUMMARY");
        summaryFilterEvents = config.getList("summary.filter.events", defaultSummaryEvents);
        apiHost = config.get("dialcode.api.host","https://qa.ekstep.in");
        dialCodeAPiEndPoint = config.get("dialcode.api.endpoint","/api/dialcode/v3/read/");
        authorizationKey = config.get("dialcode.api.authorizationkey","");
    }

    public String successTopic() {
        return successTopic;
    }

    public String failedTopic() {
        return failedTopic;
    }

    public String malformedTopic() {
        return malformedTopic;
    }

    public Integer ignorePeriodInMonths() {
        return ignorePeriodInMonths;
    }

    public String jobName() {
        return JOB_NAME;
    }

    public static String getJobFlag(String type) {
        switch (type) {
            case "device":
                return deviceDataJobFlag;
            case "user":
                return userDataJobFlag;
            case "content":
                return contentDataJobFlag;
            case "dialcode":
                return dialCodeDataJobFlag;
            case "collection":
                return collectionDataJobFlag;
            default:
                return "";
        }
    }

    public static String getDeviceLocationJobFlag() {
        return deviceDataJobFlag;
    }

    public static String getUserLocationJobFlag() {
        return userDataJobFlag;
    }

    public static String getContentLocationJobFlag() {
        return contentDataJobFlag;
    }

    public static String getCollectionLocationJobFlag() { return collectionDataJobFlag; }
    public static String getDialCodeLocationJobFlag() {
        return dialCodeDataJobFlag;
    }

    public List<String> getSummaryFilterEvents() {
        return this.summaryFilterEvents;
    }
    public String getDialCodeAPIUrl(String key){
        return apiHost.concat(dialCodeAPiEndPoint).concat(key);
    }
    public static String getAuthorizationKey(){
        return authorizationKey;
    }

}

File: data-pipeline-flink/de-normalization/src/main/scala/org/sunbird/dp/task/DenormalizationConfig.scala
package org.sunbird.dp.task

import java.util
import org.apache.flink.api.common.typeinfo.TypeInformation
import org.apache.flink.api.java.typeutils.TypeExtractor
import org.sunbird.dp.core.BaseJobConfig
import org.apache.flink.streaming.api.scala.OutputTag
import com.typesafe.config.Config
import org.sunbird.dp.domain.Event
import collection.JavaConversions._

class DenormalizationConfig(override val config: Config) extends BaseJobConfig(config) {

  private val serialVersionUID = 2905979434303791379L

  implicit val mapTypeInfo: TypeInformation[Event] = TypeExtractor.getForClass(classOf[Event])
  implicit val anyTypeInfo: TypeInformation[String] = TypeExtractor.getForClass(classOf[String])

  // Kafka Topics Configuration
  val uniqueInputTopic: String = config.getString("kafka.unique.input.topic")
  val derivedInputTopic: String = config.getString("kafka.derived.input.topic")
  val denormSuccessTopic: String = config.getString("kafka.output.success.topic")
  val failedTopic: String = config.getString("kafka.output.failed.topic")
  val metricsTopic: String = config.getString("kafka.output.metrics.topic")

  val userStore: Int = config.getInt("redis.database.userstore.id")
  val contentStore: Int = config.getInt("redis.database.contentstore.id")
  val deviceStore: Int = config.getInt("redis.database.devicestore.id")
  val dialcodeStore: Int = config.getInt("redis.database.dialcodestore.id")
  val metricsWindowSize: Int = 300;

  val deviceFields = List("country_code", "country", "state_code", "state", "city", "district_custom", "state_code_custom",
    "state_custom", "user_declared_state", "user_declared_district", "devicespec", "firstaccess")
  val contentFields = List("name", "objectType", "contentType", "mediaType", "language", "medium", "mimeType", "createdBy",
    "createdFor", "framework", "board", "subject", "status", "pkgVersion", "lastSubmittedOn", "lastUpdatedOn", "lastPublishedOn")
  val userFields = List("usertype", "grade", "language", "subject", "state", "district", "usersignintype", "userlogintype")
  val dialcodeFields = List("identifier", "channel", "batchcode", "publisher", "generated_on", "published_on", "status")
  
  val ignorePeriodInMonths:Int = if(config.hasPath("telemetry.ignore.period.months")) config.getInt("telemetry.ignore.period.months") else 3;
  val summaryFilterEvents: List[String] = if(config.hasPath("summary.filter.events")) config.getStringList("summary.filter.events").toList else List("ME_WORKFLOW_SUMMARY");
  
  val userSignInTypeDefault = if(config.hasPath("user.signin.type.default")) config.getString("user.signin.type.default") else "Anonymous";
  val userLoginInTypeDefault = if(config.hasPath("user.login.type.default")) config.getString("user.login.type.default") else "NA";

  val deviceDenormParallelism: Int = config.getInt("task.device.denorm.parallelism")
  val userDenormParallelism: Int = config.getInt("task.user.denorm.parallelism")
  val contentDenormParallelism: Int = config.getInt("task.content.denorm.parallelism")
  val locDenormParallelism: Int = config.getInt("task.loc.denorm.parallelism")
  val dialcodeDenormParallelism: Int = config.getInt("task.dialcode.denorm.parallelism")

  val WITH_LOCATION_EVENTS = "with_location_events"
  val WITH_DEVICE_EVENTS = "with_device_events"
  val WITH_USER_EVENTS = "with_user_events"
  val WITH_CONTENT_EVENTS = "with_content_events"
  val WITH_DIALCODE_EVENTS = "with_dialcode_events"
  val DENORM_EVENTS = "denorm_events"
  val JOB_METRICS = "job_metrics"

  val withLocationEventsTag: OutputTag[Event] = OutputTag[Event](WITH_LOCATION_EVENTS)
  val withDeviceEventsTag: OutputTag[Event] = OutputTag[Event](WITH_DEVICE_EVENTS)
  val withUserEventsTag: OutputTag[Event] = OutputTag[Event](WITH_USER_EVENTS)
  val withContentEventsTag: OutputTag[Event] = OutputTag[Event](WITH_CONTENT_EVENTS)
  val withDialCodeEventsTag: OutputTag[Event] = OutputTag[Event](WITH_DIALCODE_EVENTS)
  val denormEventsTag: OutputTag[Event] = OutputTag[Event](DENORM_EVENTS)
  val metricOutputTag: OutputTag[String] = OutputTag[String](JOB_METRICS)

  val jobName = "de-normalization"

}
